In [1]:
import os
import sys
sys.path.append(os.path.join(".."))

from evo_models import *

%load_ext autoreload
%autoreload 2

# Trening sieci MLP przy pomocy algorytmu ewolucyjnego

In [16]:
pop = ModelPopulation(
    # architecture=MLPArchitecture(7, [20, 10, 5], 1),
    architecture=MLPArchitecture(7, [5], 1),
    dataset_name="auto_mpg",
    population_size=100,
)
pop.show_test_loss()

Best model test loss: 24.28294138898855


In [17]:
pop.train(epochs=100)

100%|█████████████| 100/100 [00:30<00:00] , Best Loss=9.3049


In [18]:
pop.train(epochs=100)

100%|█████████████| 100/100 [00:29<00:00] , Best Loss=6.5869


In [19]:
pop.train(epochs=100)

100%|█████████████| 100/100 [00:29<00:00] , Best Loss=5.6827


In [20]:
pop.train(epochs=500)

100%|█████████████| 500/500 [02:29<00:00] , Best Loss=4.6162


In [22]:
pop.train(epochs=500, mutation_strength=0.01)

100%|█████████████| 500/500 [02:25<00:00] , Best Loss=4.4578


# Trening sieci MLP przy pomocy propagacji wstecznej

### Architektura (7, [20, 10, 5], 1)

In [38]:
model = MLP(
    architecture=MLPArchitecture(7, [20, 10, 5], 1),
    dataset_name="auto_mpg",
)

In [39]:
model.train(
    epochs=int(5000),
    learning_rate=0.001,
    optimizer="rmsprop",
    rms_beta=0.9,
    save_till_best=True,
)

Training from age 0: (best_loss_test: 6.6, loss_test: 10.65) : 100%|██████████| 5000/5000 [00:08<00:00, 610.38it/s]

--------------------
Reverting to best model at age 1242
Loss on train set: 4.49	Loss on test set: 6.6
--------------------
Model training finished at age 1242 with loss 6.6 on test set


In [41]:
model.train(
    epochs=10000,
    learning_rate=0.0001,
    optimizer="rmsprop",
    rms_beta=0.999,
    save_till_best=True,
)


Training from age 1242: (best_loss_test: 6.6, loss_test: 10.94): 100%|██████████| 10000/10000 [00:15<00:00, 630.29it/s]

--------------------
Reverting to best model at age 1242
Loss on train set: 4.49	Loss on test set: 6.6
--------------------
Model training finished at age 1242 with loss 6.6 on test set


### Architektura (7, [5], 1), czyli taka sama jak dla algorytmu ewolucyjnego

In [42]:
model2 = MLP(
    architecture=MLPArchitecture(7, [5], 1),
    dataset_name="auto_mpg",
)

In [43]:
model2.train(
    epochs=int(5000),
    learning_rate=0.001,
    optimizer="rmsprop",
    rms_beta=0.9,
    save_till_best=True,
)

Training from age 0: (best_loss_test: 6.79, loss_test: 6.82) : 100%|██████████| 5000/5000 [00:04<00:00, 1177.26it/s]  

--------------------
Reverting to best model at age 5000
Loss on train set: 5.81	Loss on test set: 6.79
--------------------
Model training finished at age 5000 with loss 6.79 on test set


In [44]:
model2.train(
    epochs=int(10000),
    learning_rate=0.001,
    optimizer="rmsprop",
    rms_beta=0.9,
    save_till_best=True,
)

Training from age 5000: (best_loss_test: 6.09, loss_test: 6.83): 100%|██████████| 10000/10000 [00:08<00:00, 1168.67it/s]

--------------------
Reverting to best model at age 10514
Loss on train set: 5.43	Loss on test set: 6.09
--------------------
Model training finished at age 10514 with loss 6.09 on test set


In [46]:
model2.train(
    epochs=int(10000),
    learning_rate=0.0001,
    optimizer="rmsprop",
    rms_beta=0.9,
    regularization="l2",
    reg_lambda=1e-4,
    save_till_best=True,
)

Training from age 10526: (best_loss_test: 6.09, loss_test: 6.58): 100%|██████████| 10000/10000 [00:08<00:00, 1136.81it/s]

--------------------
Reverting to best model at age 10526
Loss on train set: 5.43	Loss on test set: 6.09
--------------------
Model training finished at age 10526 with loss 6.09 on test set


# Wnioski
Algorytm ewolucyjny osiągnął lepszy wynik od backpropagacji , nawet wspomaganej takimi praktykami jak zastosowanie rmspropa, uczenia z momentem, regularyzacji, zmian współczynnika nauczania oraz zatrzymywania procesu nauczania przy przeuczaniu

Backprop nie był w stanie na żaden sposób zejść poniżej MSE 6.0

Algorytm ewolucyjny osiągnął ostatecznie MSE < 4.5